File: py/strings_arrays/one_away.py


def one_away(first: str, second: str):
    if len(first) == len(second):
        return _one_edit_replace(first, second)
    elif len(first) - 1 == len(second):
        return _one_edit_insert(first, second)
    elif len(first) + 1 == len(second):
        return _one_edit_insert(second, first)
    return False


def _one_edit_replace(first: str, second: str) -> bool:
    found_difference = False
    for idx, char in enumerate(first):
        if first[idx] != second[idx]:
            if found_difference:
                return False
            found_difference = True
    return True


def _one_edit_insert(first: str, second: str) -> bool:
    first_index = 0
    second_index = 0
    while second_index < len(second) and first_index < first_index:
        if first[first_index] != second[second_index]:
            if first_index != second_index:
                return False
            second_index += 1
        else:
            first_index += 1
            second_index += 1
    return True

File: java/src/com/edmundmartin/algos/stringsArrays/OneAway.java
package com.edmundmartin.algos.stringsArrays;

public class OneAway {

    public boolean oneEditAway(String first, String second) {
        if (first.length() == second.length()) {
            return oneEditReplace(first, second);
        } else if (first.length() + 1 == second.length()) {
            return oneEditInsert(first, second);
        } else if (first.length() - 1 == second.length()) {
            return oneEditInsert(second, first);
        }
        return false;
    }

    private boolean oneEditReplace(String first, String second) {
        boolean foundDifference = false;
        for (int i = 0; i < first.length(); i++) {
            if (first.charAt(i) != second.charAt(i)) {
                if (foundDifference) {
                    return false;
                }
                foundDifference = true;
            }
        }
        return true;
    }

    private boolean oneEditInsert(String first, String second) {
        int firstIndex = 0;
        int secondIndex = 0;
        while (secondIndex < second.length() && firstIndex < first.length()) {
            if (first.charAt(firstIndex) != second.charAt(secondIndex)) {
                if (firstIndex != secondIndex) {
                    return false;
                }
                secondIndex++;
            } else {
                firstIndex++;
                secondIndex++;
            }
        }
        return true;
    }
}
